app/channels/api_channel.rb

```diff
@@ -93,6 +107,33 @@ def zero_results data
     end
   end
 
+  def finish data
+    return unless scoring_open data
+
+    task = Task.find task_id
+
+    users_without_result = []
+    ActiveRecord::Base.transaction do
+      task.update! scoring_open: false
+
+      task.contest.users.find_each do |user|
+        user_result = CriterionUserResult.where user:, criterion: task.criterions
+        users_without_result << user.judge_secret unless user_result.count == task.criterions_count
+        score = user_result.sum :value
+        Result.create_or_find_by!(user:, task:).update!(score:)
+      end
+
+      raise "Users that missing some result: #{users_without_result}" if users_without_result.length.positive?
+
+      RedisLockManager.release_all client_id
```

Comment:
We should rename `release_all` to make clear that it releases all locks for a given client only. 

And create another method to release all locks for all clients.

Reply:
added